src/utils/compile.rs

```diff
@@ -0,0 +1,180 @@
+
+    slice.to_vec()
+}
+
+/// read a compiled function from bytecode, see to_bytecode for an example
+#[allow(dead_code)]
+pub fn from_bytecode(
+    context: &ContextWrapper,
+    bytecode: Vec<u8>,
+) -> Result<OwnedValueRef, ExecutionError> {
+    assert!(!bytecode.is_empty());
+    {
+        #[cfg(target_pointer_width = "64")]
+            let len = bytecode.len() as u64;
+        #[cfg(target_pointer_width = "32")]
+            let len = bytecode.len() as u32;
+
+        let buf = bytecode.as_ptr();
+        let raw = unsafe { q::JS_ReadObject(context.context, buf, len, q::JS_READ_OBJ_BYTECODE as i32) };
+
+        let func_ref = OwnedValueRef::new_dup(context, raw);
+        if func_ref.is_exception() {
+            let ex_opt = context.get_exception();
+            if let Some(ex) = ex_opt {
+                Err(ex)
+            } else {
+                Err(ExecutionError::Internal(
+                    "from_bytecode failed and could not get exception".to_string()
+                ))
+            }
+        } else {
+            Ok(func_ref)
+        }
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use crate::utils::compile::{compile, to_bytecode, from_bytecode, run_compiled_function};
+    use crate::bindings::{ContextWrapper};
+    use crate::JsValue;
+
+    #[test]
+    fn test_compile() {
+        let ctx = ContextWrapper::new(None).unwrap();
+
+        let func_res =
+            compile(
+                &ctx,
+                "{let a_tb3 = 7; let b_tb3 = 5; a_tb3 * b_tb3;}",
+                "test_func.es",
+            );
+        let func = func_res.ok().expect("func compile failed");
+        let bytecode: Vec<u8> = to_bytecode(&ctx, &func);
+        drop(func);
+        assert!(!bytecode.is_empty());
+        let func2_res = from_bytecode(&ctx, bytecode);
+        let func2 = func2_res.ok().expect("could not read bytecode");
+        let run_res = run_compiled_function(&ctx, &func2);
+        match run_res {
+            Ok(res) => {
+                assert_eq!(res.to_value().unwrap(), JsValue::Int(7 * 5));
+            }
+            Err(e) => {
+                panic!("run failed1: {}", e);
+            }
+        }
+    }
+
+    #[test]
+    fn test_bytecode() {
+        let ctx = ContextWrapper::new(None).unwrap();
+
+        let func_res = compile(
+            &ctx,
+            "let a_tb4 = 7; let b_tb4 = 5; a_tb4 * b_tb4;",
+            "test_func.es",
+        );
+        let func = func_res.ok().expect("func compile failed");
+        let bytecode: Vec<u8> = to_bytecode(&ctx, &func);
+        drop(func);
+        assert!(!bytecode.is_empty());
+        let func2_res = from_bytecode(&ctx, bytecode);
+        let func2 = func2_res.ok().expect("could not read bytecode");
+        let run_res = run_compiled_function(&ctx, &func2);
+
+        match run_res {
+            Ok(res) => {
+                assert_eq!(res.to_value().unwrap(), JsValue::Int(7 * 5));
+            }
+            Err(e) => {
```

Comment:
You can just do `run_res.expect("run failed")` here.

Reply:
is there a shorter way to do this whithout obscuring the original js error?